draft-yasskin-http-origin-signed-responses.md

```diff
@@ -321,17 +348,485 @@ frames, or could
+item from the significant request and response headers and the response payload.
+Then it categorizes each `Signature` header as "valid" or "invalid" by
+validating that header's certificate, metadata, and signature against the
+significant parts of the pair ({{signature-validity}}). This validity then
+informs higher-level protocols.
+
+The `Signature` header includes information to let a client fetch assurance a
+signed exchange is still valid, in the face of revoked certificates and
+newly-discovered vulnerabilities. This assurance can be bundled back into the
+signed exchange and forwarded to another client, which won't have to re-fetch
+this validity information for some period of time.
+
+## The Signed-Request-Headers and Signed-Response-Headers Headers ## {#signed-r-headers}
+
+Each of the `Signed-Request-Headers` and `Signed-Response-Headers` headers
+identifies an ordered list of request and response, respectively, header field
+names to include in a signature. These lists of header field names need to be
+consistent across all signatures of an exchange, to avoid forcing higher-level
+protocols to merge the header field lists of valid signatures, so they're
+provided by headers instead of being included in the `Signature` header field.
+
+`Signed-Request-Headers` and `Signed-Response-Headers` are Structured Headers as
+defined by {{!I-D.nottingham-structured-headers}}. Their values MUST be lists
+({{!I-D.nottingham-structured-headers}}, section 4.7) of lowercase strings
+({{!I-D.nottingham-structured-headers}}, section 4.2) naming HTTP headers,
+following the ordering and lowercasing rules from {{!RFC7540}}, section 8.1.2.
+
+## The Signature Header ## {#signature-header}
+
+The `Signature` header conveys a signature for an exchange and information about
+how to determine the authority of and refresh that signature.
+
+The `Signature` header is a Structured Header as defined by
+{{!I-D.nottingham-structured-headers}}. Its value MUST be a dictionary
+({{!I-D.nottingham-structured-headers}}, section 4.7).
+
+The dictionary MUST contain members named "validityUrl", "date", "expires", and
+"sig", and either "certUrl" and "certSha256" members or an "ed25519Key" member.
+The present members MUST have the following values:
+
+"certUrl"
+
+: A string ({{!I-D.nottingham-structured-headers}}, section 4.2) containing a
+  [valid URL string](https://url.spec.whatwg.org/#valid-url-string).
+
+"certSha256"
+
+: Binary content ({{!I-D.nottingham-structured-headers}}, section 4.5) holding
+  the SHA-256 hash of the first certificate found at "certUrl".
+
+"ed25519Key"
+
+: Binary content ({{!I-D.nottingham-structured-headers}}, section 4.5) holding
+  an Ed25519 public key ({{!RFC8032}}).
+
+{:#signature-validityurl} "validityUrl"
+
+: A string ({{!I-D.nottingham-structured-headers}}, section 4.2) containing a
+  [valid URL string](https://url.spec.whatwg.org/#valid-url-string).
+
+"date" and "expires
```

Comment:
missing "\""

Reply:
Also: why do we need to reinvent `cache-control` here? Or do those have a different purpose?

Reply:
9991607 now describes why I'm reinventing cache-control.